### PR TITLE
fix: texture invalid in non-live state

### PR DESCRIPTION
**Pull request by asterwyx**
Buffer update is not controlled by live property. Previous buffer is not refered, thus might be destroyed. Pends new buffer in non-live state.

### CODE REVIEW

File: src/server/qtquick/wsurfaceitem.cpp [hunk 1]
```diff
@@ -244,6 +250,13 @@ class Q_DECL_HIDDEN WSurfaceItemContentPrivate: public QQuickItemPrivate
         q->setImplicitSize(s.width(), s.height());
     }
 
+    void swapBufferIfNeeded() {
+        if (pendingBuffer) {
+            buffer.reset(pendingBuffer.release());
+            buffer->lock();
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:256]: the lock must before release to ensure the buffer is valid

File: src/server/qtquick/wsurfaceitem.cpp [hunk 2]
```diff
@@ -244,6 +250,13 @@ class Q_DECL_HIDDEN WSurfaceItemContentPrivate: public QQuickItemPrivate
         q->setImplicitSize(s.width(), s.height());
     }
 
+    void swapBufferIfNeeded() {
+        if (pendingBuffer) {
+            buffer.reset(pendingBuffer.release());
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:255]: the buffer maybe drop on release

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/wsurfaceitem.cpp [hunk 3]
```diff
@@ -244,6 +250,13 @@ class Q_DECL_HIDDEN WSurfaceItemContentPrivate: public QQuickItemPrivate
         q->setImplicitSize(s.width(), s.height());
     }
 
+    void swapBufferIfNeeded() {
+        if (pendingBuffer) {
+            pendingBuffer->lock();
```
Comment (zccrs) [inline src/server/qtquick/wsurfaceitem.cpp:255]:
````
buffer->reset(pendingBuffer.get());
buffer->lock();
pendingBuffer.reset();
````

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/wsurfaceitem.cpp (hunk 3 above)
Comment (zccrs) [reply src/server/qtquick/wsurfaceitem.cpp:255]: Maybe only need `buffer.reset(pendingbuffer.release())`

Comment (asterwyx) [reply src/server/qtquick/wsurfaceitem.cpp]: Done.

**Review by zccrs — APPROVED** [review]

**Pull request merged**